hibernate-core/src/main/java/org/hibernate/collection/internal/PersistentBag.java

```diff
@@ -99,7 +104,12 @@ public PersistentBag(SessionImplementor session, Collection coll) {
 
 	@Override
 	public boolean isWrapper(Object collection) {
-		return bag==collection;
+		if ( providedCollection == null || providedCollection == bag ) {
+			return bag == collection;
+		}
+		else {
+			return providedCollection == collection;
+		}
```

Comment:
Hibernate does not maintain consistency between `bag` and `providedCollection`. I looked at checking to see if the entries were consistent, similar to what is done by `PersistentBag#equalsSnapshot`. As you know, that can be an expensive check. 

In this case `role` is `null`, so in order to do the check, a new defaulted SPI method would need to be added that provides the role. 

I was thinking that if they were not consistent, then `IllegalStateException` should be thrown. 

I've started making these changes, but it got a bit complicated, and I'd like to get your take on this before spending more time on this.

@Steve, WDYT? Should the consistency check be done?

Reply:
In retrospect, ` || providedCollection == bag` can be dropped from the if-statement.
`

Reply:
> Hibernate does not maintain consistency between `bag` and `providedCollection`

You just added `providedCollection`, so yeah Hibernate would not be doing anything with it.